chaincode/compute_plan_worker_state.go

```diff
@@ -0,0 +1,194 @@
+	wState, err := db.GetCPWorkerState(wStateKey)
+	if err != nil {
+		return nil
+	}
+	wState.DoneCount++
+	return db.Put(wStateKey, wState)
+}
+
+// getTupleCounts returns the number of tuples in the "done" state and the total number of tuples
+// for a given compute plan
+func (cp *ComputePlan) getTupleCounts(db *LedgerDB) (doneCount int, tupleCount int, err error) {
+	for _, worker := range cp.Workers {
+		wStateKey := cp.getCPWorkerStateKey(worker)
+		wState, err := db.GetCPWorkerState(wStateKey)
+		if err != nil {
+			return doneCount, tupleCount, nil
+		}
+		doneCount += wState.DoneCount
+		tupleCount += wState.TupleCount
+	}
+	return doneCount, tupleCount, nil
+}
+
+// getCPWorkerStateKey returns the worker state key for a given compute plan and worker
+func (cp *ComputePlan) getCPWorkerStateKey(worker string) string {
+	return fmt.Sprintf("computePlan~%v~doneCountByWorker~%v", cp.Key, worker)
+}
+
+// removeAllIntermediaryModels iterates through all the worker states, and clears the lists of
+// intermediary models. It returns the concatenated list of all the intermediary models that
+// have been removed from the worker states.
+func (cp *ComputePlan) removeAllIntermediaryModels(db *LedgerDB) ([]string, error) {
+	res := []string{}
+	for _, worker := range cp.Workers {
+		wStateKey := cp.getCPWorkerStateKey(worker)
+		wState, err := db.GetCPWorkerState(wStateKey)
+		if err != nil {
+			return []string{}, err
+		}
+		res = append(res, wState.IntermediaryModelsInUse...)
+		wState.IntermediaryModelsInUse = []string{}
+
+		// clear
+		err = db.Put(wStateKey, wState)
+		if err != nil {
+			return []string{}, err
+		}
+	}
+	return res, nil
+}
+
+// getModelsInUse takes a list of model keys and checks whether these models are still in use or not.
+// It returns the initial list split into two sublists: the models in use, and the models unused.
+func getModelsInUse(db *LedgerDB, modelKeys []string) (usedModels []string, unusedModels []string, err error) {
+	usedModels = []string{}
+	unusedModels = []string{}
+	for _, modelKey := range modelKeys {
+		inUse, err := isModelInUse(db, modelKey)
+		if err != nil {
+			return []string{}, []string{}, err
+		}
+		if inUse {
+			usedModels = append(usedModels, modelKey)
+		} else {
+			unusedModels = append(unusedModels, modelKey)
+		}
+	}
+	return usedModels, unusedModels, nil
+}
+
+// isModelInUse returns true if the model with the supplied key is the
+// in-model of another training task that isn't in the "done" state.
+func isModelInUse(db *LedgerDB, modelKey string) (bool, error) {
+	keys, err := db.GetIndexKeys("tuple~modelKey~key", []string{"tuple", modelKey})
+	if err != nil {
+		return false, err
+	}
+	if len(keys) == 0 {
+		// This occurs for the keys added during the same transaction. But
+		// thoses models can just be considered in use
+		return true, nil
+	}
+	children, err := getTupleChildren(db, keys[0], true)
+	if err != nil {
+		return false, nil
```

Comment:
why you are not handling this error? is it intentional?

Reply:
Not intentional, it's a mistake

Reply:
fixed in https://github.com/SubstraFoundation/substra-chaincode/pull/138/commits/247efff256282ad69bded0a46c7f99ed422b1d0b